ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,396 @@
+//! ```
+use crate::isl::isl_range::Range;
+use crate::isl::isl_type_reference::NullabilityModifier;
+use crate::isl::version_based_isl::constraints::IslConstraint;
+use ion_rs::element::Element;
+use ion_rs::IonType;
+use std::marker::PhantomData;
+
+pub mod constraints;
+
+pub trait IslVersionTrait {}
+
+pub struct IslV1_0 {}
+impl IslVersionTrait for IslV1_0 {}
+
+// TODO: This can be a trait if we have get a new minor version of ISL
+#[derive(Debug, Clone)]
+pub struct IslV2_0 {}
+impl IslVersionTrait for IslV2_0 {}
+
+#[derive(Debug, Clone, Default, PartialEq)]
+pub struct UserReservedFields<V: IslVersionTrait> {
+    schema_header_fields: Vec<String>,
+    schema_footer_fields: Vec<String>,
+    type_fields: Vec<String>,
+    phantom: PhantomData<V>,
+}
+
+impl UserReservedFields<IslV2_0> {
+    pub fn new(
+        schema_header_fields: Vec<String>,
+        schema_footer_fields: Vec<String>,
+        type_fields: Vec<String>,
+    ) -> Self {
+        Self {
+            schema_header_fields,
+            schema_footer_fields,
+            type_fields,
+            phantom: Default::default(),
+        }
+    }
+
+    pub fn schema_header_fields(&self) -> &Vec<String> {
+        &self.schema_header_fields
+    }
+
+    pub fn schema_footer_fields(&self) -> &Vec<String> {
+        &self.schema_footer_fields
+    }
+
+    pub fn type_fields(&self) -> &Vec<String> {
+        &self.type_fields
+    }
+}
+
+/// Provides an internal representation of an schema file
+#[derive(Debug, Clone)]
+pub struct IslSchema<V: IslVersionTrait> {
+    /// Represents an id for the given ISL model
+    id: String,
+    /// Represents the user defined reserved fields
+    /// For ISL 2.0 this contains the use reserved fields that are defined within schema header,
+    /// Otherwise, it is None.
+    user_reserved_fields: Option<UserReservedFields<V>>,
+    /// Represents all the IslImports inside the schema file.
+    /// For more information: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#imports
+    imports: Vec<IslImport>,
+    /// Represents all the IslType defined in this schema file.
+    /// For more information: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#type-definitions
+    types: Vec<IslType<V>>,
+    /// Represents all the inline IslImportTypes in this schema file.
+    inline_imported_types: Vec<IslImportType>,
+    /// Represents open content as `Element`s
+    /// Note: This doesn't preserve the information about where the open content is placed in the schema file.
+    /// e.g. This method doesn't differentiate between following schemas with open content:
+    /// ```ion
+    /// $foo
+    /// $bar
+    /// type::{ name: foo, codepoint_length: 1 }
+    /// ```
+    ///
+    /// ```ion
+    /// type::{ name: foo, codepoint_length: 1 }
+    /// $foo
+    /// $bar
+    /// ```
+    open_content: Vec<Element>,
+    phantom: PhantomData<V>,
```

Comment:
I _think_ it's conventional to use `_` for unused variables/fields, and I think it's a good idea to give the phantom data a name.
```suggestion
    _version: PhantomData<V>,
```